Sources/IRGen/Expression/IRExternalCall.swift

```diff
@@ -8,13 +8,126 @@ import AST
+    guard case .functionCall(let functionCall) = externalCall.functionCall.rhs else {
+      fatalError("cannot match external call with function")
+    }
+    guard case .matchedFunction(let matchingFunction) =
+      functionContext.environment.matchFunctionCall(functionCall,
+                                                    enclosingType: functionCall.identifier.enclosingType ??
+                                                      functionContext.enclosingTypeName,
+                                                    typeStates: [],
+                                                    callerProtections: [],
+                                                    scopeContext: functionContext.scopeContext) else {
+      fatalError("cannot match function signature in external call")
+    }
+    guard let functionSelector = (matchingFunction.declaration.externalSignatureHash?.map { [$0].toHexString() }) else {
+      fatalError("cannot find function selector for function")
+    }
+
+    // Render the address of the external contract.
+    let addressExpression = IRExpression(expression: externalCall.functionCall.lhs, asLValue: false)
+      .rendered(functionContext: functionContext)
+
+    // The input stack consists of three parts:
+    // - function selector (4 bytes of Keccak-256 hash of the signature)
+    // - static data
+    // - dynamic data
+    var staticSlots: [ExpressionFragment] = []
+    var dynamicSlots: [ExpressionFragment] = []
+
+    // This could be staticSize * 32, but this loop is necessary once
+    // we have e.g. fixed-length arrays in external trait types.
+    var staticSize = 0
+    for parameterType in matchingFunction.parameterTypes {
+      switch parameterType {
+      case .solidityType:
+        staticSize += 32
+      default:
+        fatalError("cannot use non-solidity type in external call")
+      }
+    }
+
+    var dynamicSize = 0
+    for (parameterType, parameter) in zip(matchingFunction.parameterTypes, functionCall.arguments) {
+      switch parameterType {
+      case .basicType(.string):
+        // String is basic in Flint (in stack memory) but not static in Solidity
+        // Flint only supports <32 byte strings, however, because they are in
+        // stack, not in memory.
+        staticSlots.append(ExpressionFragment(pre: "", "\(staticSize + dynamicSize)"))
+        // TODO: figure out the actual length of the string at runtime
+        dynamicSlots.append(ExpressionFragment(pre: "", "32"))
+        dynamicSlots.append(IRExpression(expression: parameter.expression, asLValue: false)
+          .rendered(functionContext: functionContext))
+        dynamicSize += 32
+      case .basicType, .solidityType:
+        staticSlots.append(IRExpression(expression: parameter.expression, asLValue: false)
+          .rendered(functionContext: functionContext))
+      default:
+        fatalError("cannot use non-basic type in external call")
```

Comment:
It's probably better if this is a semantic check so we emit an appropriate diagnostic